File: 5-LongestPalindromicSubstring/5-LongestPalindromicSubstring.java
// Last updated: 2/27/2026, 5:12:17 PM
class Solution {
    public String longestPalindrome(String str) {

        if(str == null || str.length() < 1) return "";

        int start = 0;
		int end = 0;

		for(int i = 0; i < str.length() ; i++){

			int len1 = expandFromCenter(str,i - 1,i + 1);
			int len2 = expandFromCenter(str,i,i+1);

			int len = Math.max(len1,len2);

			if(len > end - start){
				start = i - ((len - 1 ) / 2);
				end = i + (len / 2);
			}

		}

		return str.substring(start,end + 1);
    }


	public  int expandFromCenter(String str, int left, int right){

        if(str == null || str.length() < 1) return 0;


		while(left >= 0 && right < str.length() && str.charAt(left) == str.charAt(right)){
			left--;
			right++;
		}

		return right - left - 1;


	}

}

File: 5-LongestPalindromicSubstring/5-LongestPalindromicSubstring.py
# Last updated: 7/12/2026, 6:26:09 PM
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        maxL = 0
        ans = ""

        for i in range(n):
            i1, j1 = self.expandFromCenter(s, i, i + 1)
            i2, j2 = self.expandFromCenter(s, i, i)

            len1 = j1 - i1 + 1
            len2 = j2 - i2 + 1

            if len1 > maxL:
                maxL = len1
                ans = s[i1:j1+1]
            
            if len2 > maxL:
                maxL = len2
                ans = s[i2:j2+1]
            
        return ans


        
    def expandFromCenter(self,s, i, j):
        n = len(s)

        while i >= 0 and j < n and s[i] == s[j]:
            i-=1
            j+=1

        return [i + 1, j - 1]
        


        
        
